README.md
# url-unshort

[![CI](https://github.com/nodeca/url-unshort/actions/workflows/ci.yml/badge.svg)](https://github.com/nodeca/url-unshort/actions/workflows/ci.yml)
[![NPM version](https://img.shields.io/npm/v/url-unshort.svg?style=flat)](https://www.npmjs.org/package/url-unshort)

> This library expands urls provided by url shortening services (see [full list](https://github.com/nodeca/url-unshort/blob/master/domains.yml)).


## Why should I use it?

It has been [argued](http://joshua.schachter.org/2009/04/on-url-shorteners) that
“shorteners are bad for the ecosystem as a whole”. In particular, if you're
running a forum or a blog, such services might cause trouble for your users:

 - such links load slower than usual (shortening services require an extra DNS
   and HTTP request)
 - it adds another point of failure (should this service go down, the links will
   die; [301works](https://archive.org/details/301works) tries to solve this,
   but it's better to avoid the issue in the first place)
 - users don't see where the link points to (tinyurl previews don't *really*
   solve this)
 - it can be used for user activity tracking
 - certain shortening services are displaying ads before redirect
 - shortening services can be malicious or be hacked so they could redirect to
   a completely different place next month

Also, short links are used to bypass the spam filters. So if you're implementing
a domain black list for your blog comments, you might want to check where all
those short links *actually* point to.


## Installation

```js
$ npm install url-unshort
```

## Basic usage

```js
const uu = require('url-unshort')()

try {
  const url = await uu.expand('http://goo.gl/HwUfwd')

  if (url) console.log('Original url is: ${url}')
  else console.log('This url can\'t be expanded')

} catch (err) {
  console.log(err);
}
```

## Retrying errors

Temporary network errors are retried automatically once (`options.request.retry=1` by default).

You may choose to retry some errors after an extended period of time using code like this:

```js
const uu = require('url-unshort')()
const { isErrorFatal } = require('url-unshort')
let tries = 0

while (true) {
  try {
    tries++
    const url = await uu.expand('http://goo.gl/HwUfwd')

    // If url is expanded, it returns string (expanded url);
    // "undefined" is returned if service is unknown
    if (url) console.log(`Original url is: ${url}`)
    else console.log("This url can't be expanded")
    break

  } catch (err) {
    // use isErrorFatal function to check if url can be retried or not
    if (isErrorFatal(err)) {
      // this url can't be expanded (e.g. 404 error)
      console.log(`Unshort error (fatal): ${err}`)
      break
    }

    // Temporary error, trying again in 10 minutes
    // (5xx errors, ECONNRESET, etc.)
    console.log(`Unshort error (retrying): ${err}`)
    if (tries >= 3) {
      console.log(`Too many errors, aborting`)
      break
    }
    await new Promise(resolve => setTimeout(resolve, 10 * 60 * 1000))
  }
}
```


## API

### Creating an instance

When you create an instance, you can pass an options object to fine-tune unshortener behavior.

```js
const uu = require('url-unshort')({
  nesting: 3,
  cache: {
    get: async key => {},
    set: async (key, value) => {}
  }
});
```

Available options are:

- **nesting** (Number, default: `3`) - stop resolving urls
  when `nesting` amount of redirects is reached.

  It happens if one shortening service refers to a link belonging to
  another shortening service which in turn points to yet another one
  and so on.

  If this limit is reached, `expand()` will return an error.

- **cache** (Object) - set a custom cache implementation (e.g. if you wish
  to store urls in Redis).

  You need to specify 2 promise-based functions, `set(key, value)` & `get(key)`.

- **request** (Object) - default options for
  [got](https://github.com/sindresorhus/got) in `.request()` method. Can be
  used to set custom `User-Agent` and other headers.


### uu.expand(url) -> Promise

Expand an URL supplied. If we don't know how to expand it, returns `null`.

```js
const uu = require('url-unshort')();

try {
  const url = await uu.expand('http://goo.gl/HwUfwd')

  if (url) console.log('Original url is: ${url}')
  // no shortening service or an unknown one is used
  else console.log('This url can\'t be expanded')

} catch (err) {
  console.log(err)
}
```

### uu.add(domain [, options])

Add a new url shortening service (domain name or an array of them) to the white
list of domains we know how to expand.

```js
uu.add([ 'tinyurl.com', 'bit.ly' ])
```

The default behavior will be to follow the URL with a HEAD request and check
the status code. If it's `3xx`, return the `Location` header. You can override
this behavior by supplying your own function in options.

Options:

- **aliases** (Array) - Optional. List of alternate domaine names, if exist.
- **match** (String|RegExp) - Optional. Custom regexp to use for URL match.
  For example, if you need to match wildcard prefixes or country-specific
  suffixes. If used with `validate`, then regexp may be not precise, only to
  filter out noise. If `match` not passed, then exact value auto-generated from
  `domain` & `aliases`.
- **validate** (Function) - Optional. Does exact URL check, when complex logic
  required and regexp is not enouth (when `match` is only preliminary). See
  `./lib/providers/*` for example.
- **fetch**  (Function) - Optional. Specifies custom function to retrieve expanded
  url, see `./lib/providers/*` for examples. If not set - default method used
  (it checks 30X redirect codes & `<meta http-equiv="refresh" content='...'>`
  in HTML).
- **link_selector** (String) - Optional. Some sites may return HTML pages instead
  of 302 redirects. This option allows use jquery-like selector to extract
  `<a href="...">` value.

Example:

```js
const uu = require('url-unshort')()

uu.add('notlong.com', {
  match: '^(https?:)//[a-zA-Z0-9_-]+[.]notlong[.]com/'
})

uu.add('tw.gs', {
  link_selector: '#lurllink > a'
})
```

### uu.remove(domain)

(String|Array|Undefined). Opposite to `.add()`. Remove selected domains from
instance config. If no params passed - remove everything.


## Security considerations

Only `http` and `https` protocols are allowed in the output. Browsers technically
support redirects to other protocols (like `ftp` or `magnet`), but most url
shortening services limit redirects to `http` and `https` anyway. In case
service redirects to an unknown protocol, `expand()` will return an error.

`expand()` function returns url from the url shortening **as is** without any
escaping or even ensuring that the url is valid. If you want to guarantee a
valid url as an output, you're encouraged to re-encode it like this:

```js
var URL = require('url');

url = await uu.expand('http://goo.gl/HwUfwd')

if (url) url = URL.format(URL.parse(url, null, true))

console.log(url));
```

## License

[MIT](https://raw.github.com/nodeca/url-unshort/master/LICENSE)


CHANGELOG.md
6.1.0 / 2022-10-22
------------------

- Deps bump.
- Deprecated `soo.gd` & `korta.nu`.
- Fixed `vk.cc` & `vurl.com`.


6.0.0 / 2022-05-17
------------------

- Cleanup deprecated redirectors. Move deprecation info to separate file.
- node.js v14+ required.
- Renamed option `select` => `link_selector`.
- Added method `.remove()` method.
- Add `isErrorFatal` helper.
- Deps bump.


5.0.0 / 2017-06-08
------------------

- Switch to native async/await (need nodejs 7.+)
- Drop callbacks support.


4.1.0 / 2017-06-08
------------------

- Maintenance, deps bump. `got` 6.x -> 7.x. `got` timeouts may work a bit
  different but should affect result.


4.0.0 / 2016-12-08
------------------

- Move request options to `options.request`.
- Update default User-Agent string.
- Deprecate `error.status` (use `error.statusCode`).
- Add more info (code) to error messages.
- flic.kr should use `.request()` method.
- Increase default request timeout to 30 seconds.


3.1.0 / 2016-12-05
------------------

 - `err.status` -> `err.statusCode` (old `err.status` still exists for backward
   compatibility, but will be deprecated).


3.0.0 / 2016-11-27
------------------

- Rewrite internals to promises (including .require() / cache.get() /
  cache.set()).
- Drop old node.js support, now v4.+ required.


2.1.0 / 2016-07-15
------------------

- Added `google.*/url` unshortening.
- Reenabled some glitching services.
- Added incident dates to default config for tracking progress in future.


2.0.0 / 2016-05-24
------------------

- Added Promise support in `.expand` method.
- Services cleanup.


1.1.3 / 2016-01-17
------------------

- Maintenance: deps update.


1.1.2 / 2015-12-07
------------------

- Enchanced error info with `code` & `status` properties.


1.1.1 / 2015-11-27
------------------

- Improved cache use for edge case with empty result.


1.1.0 / 2015-11-25
------------------

- Optimized cache use. Store data only if fetch happened.
- Increased request timeout to 10 seconds.


1.0.1 / 2015-10-28
------------------

- Added `vk.com/away.php` support.


1.0.0 / 2015-08-16
------------------

- First release.


lib/index.js
// Main class
//

'use strict'

const $ = require('cheerio/lib/slim').load('')
const read = require('fs').readFileSync
const yaml = require('js-yaml')
const path = require('path')
const punycode = require('punycode/')
const got = require('got')
const escapeRe = require('escape-string-regexp')
const merge = require('lodash.merge')
const URL = require('url').URL
const UnshortError = require('./error')
const pkg = require('../package.json')

const config = yaml.load(read(path.join(__dirname, '..', 'domains.yml'), 'utf8'))

const defaultAgent = `${pkg.name}/${pkg.version} (+https://github.com/nodeca/url-unshort)`

const defaultOptions = {
  timeout: 30 * 1000,
  retry: 1,
  followRedirect: false, // redirects are handled manually
  headers: {
    'User-Agent': defaultAgent
  }
}

const customProviders = require('./providers/index')

// Create an unshortener instance
//
// options:
//  - cache   (Object) - cache instance
//    - get(key) -> Promise
//    - set(key, value)  -> Promise
//  - nesting (Number) - max amount of redirects to follow, default: `3`
//  - request (Object) - default options for `got` in `.request()` method
//
function Unshort (options = {}) {
  if (!(this instanceof Unshort)) return new Unshort(options)

  this._options = merge({}, defaultOptions, options.request || {})

  // config data with compiled regexps and fetch functions attached
  this._sites = []
  this._compiled_sites = []

  this.cache = options.cache || {
    get: async () => {},
    set: async () => {}
  }

  this.nesting = options.nesting || 3

  // Regexp that matches links to all the known services, it is used
  // to determine whether url should be processed at all or not.
  //
  // Initialized to regexp that matches nothing, it gets overwritten
  // when domains are added.
  //
  this._matchAllRE = /(?!)/

  // Merge config data & custom providers
  config.forEach(site => {
    let domain, options

    if (typeof site === 'string') {
      [domain, options] = [site, {}]
    } else {
      [domain, options] = Object.entries(site)[0]
    }

    if (customProviders[domain]) {
      Object.assign(options, customProviders[domain])
    }

    this.add(domain, options)
  })
}

// Remove previously added domain.
//
//  - domain (String|Array) - list of domain names (leave undefined to drop all)
//
Unshort.prototype.remove = function (domain) {
  if (!domain) {
    this._sites.length = 0
  } else if (!Array.isArray(domain)) {
    this._sites = this._sites.filter(s => s.id !== domain)
  } else {
    for (const d of domain) {
      this._sites = this._sites.filter(s => s.id !== d)
    }
  }

  this._compile()
}

// Add a domain name to the list of known domains
//
//  - domain (String|Array) - list of domain names
//  - options (Object)       - options for these domains
//    - link_selector (String)  - jquery-like selector to retrieve url with
//    - match (String|RegExp)   - custom regexp to use to match this domain
//    - fetch (Function)         - custom function to retrieve expanded url
//
Unshort.prototype.add = function (domain, options = {}) {
  if (Array.isArray(domain)) {
    for (const d of domain) {
      this._sites.push(Object.assign({ id: d }, options))
    }
  } else {
    this._sites.push(Object.assign({ id: domain }, options))
  }

  this._compile()
}

// Normalize site data:
//
// - create default handlers if not exist
// - build `match` regexp, depending on other fields
//
// Returns normalized object, suitable for unified processing.
//
Unshort.prototype._compileSingle = function (site) {
  // Prepare list of all the domain names, including aliases
  // and punycode variations
  let dList = [site.id].concat(site.aliases || [])

  // create variations + make unique
  dList = Array.from(new Set(
    dList.map(punycode.toASCII).concat(dList.map(punycode.toUnicode))
  ))

  let match

  if (site.match) {
    // regexp is specified by a user
    match = typeof site.match === 'string'
      ? new RegExp(site.match, 'i')
      : site.match
  } else {
    // regexp is auto-generated out of domain list
    match = new RegExp(
      `^(https?:)?//(www[.])?(${dList.map(escapeRe).join('|')})/`,
      'i'
    )
  }

  return Object.assign(
    { fetch: this._defaultFetch, validate: () => true },
    site,
    { match }
  )
}

// Rebuild all regexps & default handlers for fast run
Unshort.prototype._compile = function () {
  this._compiled_sites.length = 0

  for (const site of this._sites) {
    this._compiled_sites.push(this._compileSingle(site))
  }

  // Create global search regexp
  this._matchAllRE = new RegExp(
    this._compiled_sites.map(cs => cs.match.source).join('|'),
    'i'
  )
}

// Internal method to perform an http(s) request, it's supposed to be used
// in fetchers. You can override it with custom implementation (for example,
// if you want to avoid http requests at all and use cache only, you can
// replace this with a stub).
//
Unshort.prototype.request = function (url, options) {
  const opts = merge({}, this._options, options || {})

  return got(url, opts).catch(err => {
    let statusCode = err.statusCode

    if (err.code === 'ERR_NON_2XX_3XX_RESPONSE' && err.response) {
      // https://github.com/sindresorhus/got/blob/main/documentation/8-errors.md
      statusCode = err.response.statusCode
    }

    throw new UnshortError(
      `Remote server error, code ${err.code}, statusCode ${statusCode}`,
      'EHTTP',
      statusCode)
  })
}

// Expand an URL
//
//  - url      (String)   - url to expand
//
Unshort.prototype.expand = function (url) {
  return this._expand(url)
}

// Internal method that expands url recursively up to `nesting` times,
// on each execution it parses input url and calls a fetcher of the
// matching domain.
//
Unshort.prototype._expand = async function (origUrl) {
  if (origUrl.startsWith('//')) {
    try {
      /* eslint-disable no-new */
      new URL(origUrl)
    } catch (e) {
      try {
        // set protocol for relative links like `//example.com`
        new URL('http:' + origUrl)
        origUrl = 'http:' + origUrl
      } catch {}
    }
  }

  let url = origUrl
  let shouldCache = false
  let nestingLeft = this.nesting

  for (; nestingLeft >= 0; nestingLeft--) {
    let hash = ''

    //
    // Normalize url & pre-validate
    //

    let u = new URL(url)

    // user-submitted url has weird protocol, just return `null` in this case
    if (u.protocol !== 'http:' && u.protocol !== 'https:') break

    if (u.hash) {
      // Copying browser-like behavior here: if we're not redirected to a hash,
      // but original url has one, set it as a final hash.
      hash = u.hash
      u.hash = ''
    }

    const urlNormalized = u.toString()

    //
    // At top level try cache first. On recursive calls skip cache.
    // !! Cache should be probed even for disabled services, to resolve old links.
    //
    let result

    if (nestingLeft === this.nesting) {
      result = await this.cache.get(urlNormalized)

      // If cache exists - use it.
      if (result || result === null) {
        // forward hash if needed
        if (hash && result) {
          u = new URL(result)
          u.hash = u.hash || hash
          result = u.toString()
        }

        return result
      }
    }

    //
    // First pass validation (quick).
    //

    if (!this._matchAllRE.test(urlNormalized)) break

    // Something found - run additional checks.

    const siteConfig = this._compiled_sites.find(cs => cs.match.exec(urlNormalized))

    if (!siteConfig || !siteConfig.validate(urlNormalized)) break

    // Valid redirector => should cache result
    shouldCache = true

    result = await siteConfig.fetch.call(this, urlNormalized, siteConfig)

    // If unshortener has persistent fail - stop.
    if (!result) break

    // Parse and check url
    //
    try {
      u = new URL(result)
    } catch (e) {
      if (e instanceof TypeError && e.message === 'Invalid URL') {
        throw new UnshortError('Redirected to an invalid location', 'EBADREDIRECT')
      }

      throw e
    }

    if (u.protocol !== 'http:' && u.protocol !== 'https:') {
      // Accept:
      //
      //  - http:// protocol (e.g. http://example.org/)
      //  - https:// protocol (e.g. https://example.org/)
      //
      // Restriction is done for security reasons. Even though browsers
      // can redirect anywhere, most shorteners have similar restrictions.
      //
      throw new UnshortError('Redirected to an invalid location', 'EBADREDIRECT')
    }

    // restore hash if needed
    if (hash && !u.hash) {
      u.hash = hash
      result = u.toString()
    }

    url = result
  }

  if (nestingLeft < 0) {
    throw new UnshortError('Too many redirects', 'EBADREDIRECT')
  }

  const result = (url !== origUrl) ? url : null

  if (shouldCache) {
    // Cache result.
    // !! use normalized original URL for cache key.
    const uo = new URL(origUrl)

    uo.hash = ''

    await this.cache.set(uo.toString(), result)
  }

  return result
}

Unshort.prototype._isRedirect = function (code) {
  return [301, 302, 303, 307, 308].includes(code)
}

// Default fetcher, it requests an url and retrieves url it redirects to
// using following data sources:
//
//  - "Location" header if response code is 3xx
//  - <meta http-equiv="refresh"> meta tag
//  - $(selector).attr('href, src') if selector is specified
//
Unshort.prototype._defaultFetch = async function (url, options) {
  let res

  try {
    res = await this.request(url)
  } catch (e) {
    if (e.statusCode >= 400 && e.statusCode < 500) return null
    throw e
  }

  if (this._isRedirect(res.statusCode)) {
    return res.headers.location ? res.headers.location.trim() : null
  }

  if (res.statusCode >= 200 && res.statusCode < 300) {
    if (!res.headers['content-type'] ||
        res.headers['content-type'].split(';')[0].trim() !== 'text/html') {
      return null
    }

    const body = String(res.body)

    if (options.link_selector) {
      // try to lookup selector if it's defined in the config
      const el = $(body).find(options.link_selector)
      const result = el.attr('href')

      if (result) return result.trim()
    }

    // try <meta http-equiv="refresh" content="..."> tag
    let refresh = $(body)
      .find('meta[http-equiv="refresh"]')
      .attr('content')

    if (!refresh) return null

    // parse meta-tag and remove timeout,
    // refresh at this point is like `0.5; url=http://example.org`
    refresh = refresh.replace(/^[^;]+;\s*url=/i, '').trim()

    return refresh
  }

  throw new UnshortError(
    `Remote server error, code ${res.code}, statusCode ${res.statusCode}`,
    'EHTTP',
    res.statusCode)
}

module.exports = Unshort
